packages/hardhat-core/src/internal/hardhat-network/provider/fork/ForkBlockchain.ts

```diff
@@ -254,13 +261,17 @@ export class ForkBlockchain
     }
 
     const common = this._common.copy();
-    // We set the common's hardfork to Berlin if the remote block doesn't have
-    // EIP-1559 activated. The reason for this is that ethereumjs throws if we
-    // have a base fee for an older hardfork, and set a default one for London.
-    if (rpcBlock.baseFeePerGas !== undefined) {
-      common.setHardfork("london"); // TODO: consider changing this to "latest hardfork"
-    } else {
+    // We set the common's hardfork depending on the remote block fields, to
+    // prevent ethereumjs from throwing if unsupported fields are passed.
+    // We use "berlin" for pre-EIP-1559 blocks (blocks without baseFeePerGas),
+    // "merge" for blocks that have baseFeePerGas but not withdrawals,
+    // and "shanghai" for blocks with withdrawals
+    if (rpcBlock.baseFeePerGas === undefined) {
       common.setHardfork("berlin");
+    } else if (rpcBlock.withdrawals === undefined) {
+      common.setHardfork("merge");
+    } else {
+      common.setHardfork("shanghai");
```

Comment:
This is an important and risky change. When processing a remote block, we try to set the "right" hardfork that will accept the fields of the remote block. This doesn't mean that the hardfork exactly matches the hardfork enabled in the remote block, but that it will be "compatible".

Reply:
I think it should be fine, as we were already setting incorrect HFs.